src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs

```diff
@@ -0,0 +1,225 @@
+using System.Collections.ObjectModel;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Workspaces
+{
+    [Export(typeof(ITextBufferVisibilityTracker))]
+    internal sealed class CocoaTextBufferVisibilityTracker : ITextBufferVisibilityTracker
+    {
+        private readonly ITextBufferAssociatedViewService _associatedViewService;
+        private readonly IThreadingContext _threadingContext;
+
+        private readonly Dictionary<ITextBuffer, VisibleTrackerData> _subjectBufferToCallbacks = new();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CocoaTextBufferVisibilityTracker(
+            ITextBufferAssociatedViewService associatedViewService,
+            IThreadingContext threadingContext)
+        {
+            _associatedViewService = associatedViewService;
+            _threadingContext = threadingContext;
+
+            associatedViewService.SubjectBuffersConnected += AssociatedViewService_SubjectBuffersConnected;
+            associatedViewService.SubjectBuffersDisconnected += AssociatedViewService_SubjectBuffersDisconnected;
+        }
+
+        private void AssociatedViewService_SubjectBuffersConnected(object sender, SubjectBuffersConnectedEventArgs e)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+            UpdateAllAssociatedViews(e.SubjectBuffers);
+        }
+
+        private void AssociatedViewService_SubjectBuffersDisconnected(object sender, SubjectBuffersConnectedEventArgs e)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+            UpdateAllAssociatedViews(e.SubjectBuffers);
+        }
+
+        private void UpdateAllAssociatedViews(ReadOnlyCollection<ITextBuffer> subjectBuffers)
+        {
+            // Whenever views get attached/detached from buffers, make sure we're hooked up to the appropriate events
+            // for them.
+            foreach (var buffer in subjectBuffers)
+            {
+                if (_subjectBufferToCallbacks.TryGetValue(buffer, out var data))
+                    data.UpdateAssociatedViews();
+            }
+        }
+
+        public bool IsVisible(ITextBuffer subjectBuffer)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            var views = _associatedViewService.GetAssociatedTextViews(subjectBuffer).ToImmutableArrayOrEmpty();
+
+            // If we don't have any views at all, then assume the buffer is visible.
+            if (views.Length == 0)
+                return true;
+
+            // if any of the views were *not* wpf text views, assume the buffer is visible.  We don't know how to
```

Comment:
"wpf"

Reply:
Seems like I missed one :p It's copy pasted from the WPF impl. Want me to do a split with shared code and partial classes? I imagine that makes things simpler.

Reply:
yes.  or just have a generic base type that does all the logic and is parameterized on TTextView, then subclass with the wpf and cocoa impls.

Reply:
Yeah, that sounds good. Either works for me.

Reply:
```suggestion
            // if any of the views were *not* cocoa text views, assume the buffer is visible.  We don't know how to
```